doc/developer/design/20220404_metadata_storage.md

```diff
@@ -0,0 +1,304 @@
+
+## Discussion
+
+### Current state of affairs
+
+Materialize Platform currently stores metadata in SQLite databases on an EBS
+volume. The hard requirements immediately rule this situation out as a viable
+long term option ([details][sqlite-problems]).
+
+### Abstraction
+
+`materialized` is moving towards storing all its metadata via the [`Stash`] API,
+which abstracts over any particular database's API. This API is intentionally
+kept simple enough that `Stash` can, in principle, be backed by any of the
+following:
+
+  * A SQL database with strictly serializable transactions
+  * A key–value store with strictly serializable transactions
+  * Materialize's own STORAGE layer
+
+There has been some discussion of introducing a networked service to mediate
+between `materialized` and the database. The idea is that this service would
+decouple `materialized` from the database, allowing for changes to the database
+interactions to be deployed on a separate schedule from `materialized`. However,
+this network service would introduce at least several milliseconds of latency,
+additional operational burden, and additional backwards compatibility surface
+area. I think `Stash` provides the same decoupling benefits without any of the
+other downsides. `materialized` is code that's entirely under our control; we
+can deploy any necessary changes to `materialized` with whatever frequency is
+necessary.
+
+### Options for evaluation
+
+* DynamoDB
+
+  Fatal flaw: transactions have an observed median latency of 15ms, with tail
+  latencies of 100ms+.
+
+* A PostgreSQL RDS/Aurora instance per customer.
+
+  Fatal flaw: provisioning a new RDS instance takes tens of minutes. Far too
+  slow to support user onboarding.
+
+* A PostgreSQL RDS/Aurora instance per cloud provider region.
+
+  Open concern: PostgreSQL has an extremely strict limit on the maximum number
+  of connections, and we'll need one connection per environment. Connection
+  poolers like PgBouncer only limit this to the extent that the environments
+  are not constantly issuing queries.
+
+  Open concern: PostgreSQL is not particularly performant in `SERIALIZABLE`
+  mode. Perhaps this won't impact us because we only expect transactions to
+  conflict in the unusual and transient case of having two live controller
+  processes.
+
+  Open concern: a single Aurora instance has a limit of 5000 connections.
+
+* A CockroachDB cluster per cloud provider region.
+
+  Open concern: CockroachDB does not provide strict serializability. How does
+  the anomaly that CockroachDB permits (termed a "causal reverse") impact our
+  ability to provide strict serializability?
+
+  Open concern: CockroachDB does not provide single-key linearizability in the
+  case of clock skew. Empirically: clock skew happens!
+
+* A FoundationDB cluster per cloud provider region.
+
+  Open concern: there are no vendors that offer enterprise support contracts.
+
+  Open soncern: there is a 10MB limit on write transactions.
+
```

Comment:
Are there other databases (or similar) we've contemplated here? I have a slightly heretical proposal: what about a Redis (Elasticache) per region? 

### Must haves

1. Resilience to single availability zone outages
   -   Multi-AZ deployments are supported by Elasticache


1. Strong consistency
   -   As long as you&rsquo;re going through a single writer, Redis is strongly consistent

1. Millisecond latency

   -   Redis has in my experience lived up to this easily

1. Easy scalability

   -   We can go up to 400+ GiB vertically with AWS, which gives us room for months if not years

1. Compatibility with single-binary / BYOC

   -   A single binary could use something like [redisless](https://github.com/Qovery/RedisLess)
   -   Other clouds have Redis-compatible services


1. Environment schema isolation

   -   As a key-value store, we could design our application-level schema flexibly: each environment could have a separate keyspace which is not affected by others


1. Security sandboxing

   -   An Elasticache per customer might work in this case: Elasticache instances have taken in my experience <5 minutes, and this would give us strong isolation.
    -   Redis has ACLs and Elasticache does support some form of RBAC, too


### Nice to haves


1. Consistency across clouds
   -   Check


1. Ability to run in test environments

   -   Check, with a local Redis


1. Environment performance isolation

   -   Check if we do an instance per customer, no worse than Postgres otherwise


1. Migration plan to `STORAGE`

   -   This I&rsquo;m less sure of!

Reply:
Huh... I think ... I think this roughly all checks out, except I have the question above about manual failover. How do we make manual failover safe? IIUC, Redis with asynchronous replication will for sure lose a few seconds of data when you failover. And if you run with only one replica, you lose the ability to fail over at all. That's no worse than the current setup of SQLite on EBS, and possibly worth exploring if y'all think it'll be easier to manage Redis than SQLite. 

Btw, 5m of provisioning time worries me. One of the most loved features of our current cloud product is that we can turn on a new deployment in less than 60s. I'd love to similarly be able to turn on an environment in Materialize Platform in less than 60s. Makes for a much sweeter onboarding experience. We can solve that by pre-allocating Elasticache instances, but more code to write and infra to manage if we do that! All this is probably moot, though; I think ACLs give us exactly the security sandboxing that we'd need.